xain/grpc/participant.py

```diff
@@ -65,34 +86,163 @@ def end_training(
+    cid = 0
+    xy_train = xy_train_partitions[FLAGS.partition_iden]
+    return Participant(
+        cid,
+        model_provider,
+        xy_train,
+        xy_val,
+        num_classes=10,
+        batch_size=FLAGS.batch_size,
     )
-    heartbeat_thread.start()
 
-    try:
-        # never returns unless there is an exception
-        heartbeat_thread.join()
-    except KeyboardInterrupt:
-        terminate_event.set()
-        channel.close()
+
+def training_round(channel, participant: Participant):
+    theta, epochs, base = start_training(channel)
+    # training:
+    theta_n, his, _dict = participant.train_round(theta, epochs, base)
+    # NOTE _dict is the opt_config - ignore for now
+    met = participant.metrics()
+    end_training(channel, theta_n, his, met)
+
+
+def main(_argv):
+    print(f"model_name: {FLAGS.model_name}")
+    print(f"dataset_name: {FLAGS.dataset_name}")
+    print(f"batch_size: {FLAGS.batch_size}")
+    print(f"partition_iden: {FLAGS.partition_iden}")
+    go(init_participant())
 
 
 if __name__ == "__main__":
-    run()
+    flags.mark_flag_as_required("model_name")
+    flags.mark_flag_as_required("dataset_name")
+    flags.mark_flag_as_required("batch_size")
+    flags.mark_flag_as_required("partition_iden")
+    app.run(main)
+
+
+class StateRecord:
+    def __init__(self):
+        self.cv = threading.Condition()
+        self.round = 0
+        self.state = ParState.WAITING_FOR_SELECTION
+
+    def lookup(self):
+        with self.cv:
+            return self.state, self.round
+
+    def update(self, state):
+        with self.cv:
+            self.state = state
+            self.cv.notify()
+
+    def wait_until_selected_or_done(self):
+        with self.cv:
+            self.cv.wait_for(lambda: self.state in {ParState.TRAINING, ParState.DONE})
+            # which one was it?
+            return self.state
+
+    def wait_until_next_round(self):
+        with self.cv:
+            self.cv.wait_for(
+                lambda: self.state
+                in {ParState.TRAINING, ParState.WAITING_FOR_SELECTION}
+            )
+            # which one was it?
+            return self.state
+
+
+# updates st
+def transit(st, beat_reply):
+    msg, r = beat_reply.state, beat_reply.round
+    with st.cv:
+        if st.state == ParState.WAITING_FOR_SELECTION:
+            if msg == coordinator_pb2.State.ROUND:
+                st.state = ParState.TRAINING
+                st.round = r
+                st.cv.notify()
+            elif msg == coordinator_pb2.State.FINISHED:
+                st.state = ParState.DONE
+                st.cv.notify()
+        elif st.state == ParState.POST_TRAINING:
+            if msg == coordinator_pb2.State.STANDBY:
+                # not selected
+                st.state = ParState.WAITING_FOR_SELECTION
+                # prob ok to keep st.round as it is
+                st.cv.notify()
+            elif msg == coordinator_pb2.State.ROUND and r == st.round + 1:
```

Comment:
I think that here we should check if `r > st.round`. That way the participant is not required to participate in every round. Makes it a bit more flexible

Reply:
Sensible suggestion but it's more flexible than it might look. There's no selection yet in the Coordinator, but let's imagine it's communicated in perhaps the simplest way: it advertises `ROUND i+1` to the selected, and `STANDBY` to those not. Then a Participant in `POST_TRAINING i` expects just one of those 2 messages - if it's not required in round i+1 it should go back to `WAITING_FOR_SELECTION` (the case above). So a message with `r > st.round + 1` would be unexpected.

Reply:
The case that I was thinking of is if let's the participant was selected for ROUND 1 and skipped for ROUND 2 and then selected again for ROUND 3. Would this work?

Reply:
It should. The participant is selected via the heartbeat `ROUND 1` (thereby transitioning `WAITING_FOR_SELECTION` -> `TRAINING 1`). Once training is finished, it waits (`TRAINING 1` -> `POST_TRAINING 1`). At some point it realises from the heartbeat `STANDBY` that it's not required for round 2 and so it's back to more waiting (`POST_TRAINING 1` -> `WAITING_FOR_SELECTION`). Then round 2 is over, and `ROUND 3` signals it back into action (`WAITING_FOR_SELECTION` -> `TRAINING 3`). So on.